### ISSUE TITLE
Just a question.

### ISSUE THREAD

**Issue by PaulScemama** [deepmind/conformal_training#2]
Hi David,

First of all thank you for creating such a readable research codebase. It's definitely inspiring to see good and readable code in a research setting. 

## My question
I've been using your adaptive prediction sets (APS) function to do conformal prediction on a model I trained on fashion mnist.

My workflow is as follows:

```python
calibration_inputs: jnp.ndarray
calibration_outputs: jnp.ndarray

# This is a LeNet 5-layer network.
trained_model = model.load_state("path/to/model")

# NOTE: my trained model is a posterior predictive where the posterior is
# over the weights. So `.predict()` means taking 30 samples (30 forward passes)
# and then the prediction is the proportion of each class for those 30 samples.
calibration_predictions = trained_model.predict(calibration_inputs, predictive_samples=30)
calibration_accuracy = compute_accuracy(probabilities=calibration_predictions, labels=calibration_outputs)


tau = calibrate_raps(probabilities=calibration_predictions, labels=calibration_outputs, alpha=0.1)
calibration_conformal_sets = predict_raps(probabilities=calibration_predictions, tau=tau)

calibration_coverage = compute_coverage(confidence_sets=calibration_conformal_sets, labels=calibration_outputs)


print(calibration_accuracy)
# 0.875...
print(calibration_coverage)
# 0.999...
```

So what I'm seeing is that even though I set `alpha=0.1` which indicates I want ~90% coverage, I keep getting "over-coverage". And this is the coverage on the calibration set no less. In most papers I've seen that evaluate APS in some way (usually as a baseline), their coverage (albeit on the test set) coincides with approximately 1-`alpha`. 

Do you think I have a bug in my code? Or is this a symptom of APS being a generally overconservative conformal procedure (which I've heard is the case, hence RAPS)? 

Last note (and reminder) is that the model is "Bayesian" in that its a Laplace Approximation to the posterior over the weights of the network. Perhaps its already calibrated outputs being used as inputs into the APS procedure has an effect? 

Thanks!

**Comment by PaulScemama** [comment]
And just to follow up, I just went with the `THR` (threshold) conformal procedure and I'm getting the proper coverage...where what I mean by proper is that I set `alpha=0.1` and coverage is $\approx 90$%. This makes me think it is the issue of Adaptive Prediction Sets (APS).

**Comment by davidstutz** [comment]
Sorry for the late answer on this.

AFAIK it usually should not be over-conservative if the scores are all distinct. Can you check whether you have a lot of very similar scores across examples? In this case you would want to set the the `rng` argument in `predict_raps`. Let me know if that helps.

**Comment by PaulScemama** [comment]
No worries! @davidstutz. Yes that seemed to be the case and setting the `rng` has solved the issue for the most part. Can I ask what exactly is going on when we set the `rng` argument in both `calibrate_raps` and `predict_raps`? It seems like you subtract random uniform(0,1) samples from the scores? 

Thanks!

**Comment by davidstutz** [comment]
That's essentially what is going on. If you check Theorem 1 in the APS paper [1], you will see that for coverage to be tight (i.e., having an upper bound on coverage), we need distinct conformity scores. This additional uniform random variable takes care of having distinct conformity scores during calibration in `calibrate_raps`. Since the score at test time needs to be the same, it also needs to be added in `predict_raps`.

[1] https://proceedings.neurips.cc/paper/2020/file/244edd7e85dc81602b7615cd705545f5-Paper.pdf

**Comment by PaulScemama** [comment]
Ah that's what I was missing. Thanks again, this was very helpful!
